Modules/_ssl.c

```diff
@@ -6241,6 +6338,66 @@ _ssl_get_default_verify_paths_impl(PyObject *module)
     return NULL;
 }
 
+/*[clinic input]
+@critical_section
```

Comment:
Do we really need a critical section?

Reply:
I'm not sure. There is a memory allocation involved here in OpenSSL, and I don't know offhand if that's thread-safe. Also, iI think t's possible to dynamically enable/disable providers, which could end up affecting which algorithms show up in the returned list. So, I decided to include it for safety.

Reply:
AFAIR, critical sections are for protecting CPython code, not external code. The critical section is applied to the module object here, namely `ssl`, but since `ssl` itself is not modified concurrently by this call, we don't need a critical section (the OpenSSL call doesn't depend on an SSL context; for contexts the critical section is needed because you can modify the underlying context between the call to the Python API and the OpenSSL API (e.g., one thread does ctx.get_value() and this calls ssl_get_value(ctx->attr) while another thread does ctx.set_attr(other), and you'll get a race condition because ctx->attr was concurrently modifed).

Am I right @ZeroIntensity?

Reply:
Ok, thanks!

For completeness, I should mention that there is an argument to `SSL_get1_builtin_sigalgs` of type OSSL_LIB_CTX *, but Python doesn't appear to ever create custom library contexts so the code always passes NULL in here. Given that multiple threads could be referencing the same default library context, though, does that change anything? This call would be read-only on that library context's state, but I wonder if there could be other threads which change state in the default library context.